### PR TITLE
Add course

### LINKED ISSUE #63: Add course to contact
As an NUS computing student, I wish to add a course taken/taking to each of my contacts, so that I can store useful knowledge about the person’s courses taken for future reference.

### PR DESCRIPTION
**Pull request by Eola-Z**

### CODE REVIEW

File: src/main/java/networkbook/model/person/Course.java [hunk 1]
```diff
@@ -3,45 +3,145 @@
...
+        for (String word : test.split(" ")) {
+            if ((word.equals(""))) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Return true if the given string is a valid date.
```
Comment (xenosf) [inline src/main/java/networkbook/model/person/Course.java:95]:
```suggestion
     * Returns true if the given string is a valid date.
```

File: src/main/java/networkbook/model/person/Course.java [hunk 2]
```diff
@@ -3,45 +3,145 @@
...
+        checkArgument(isValidCourse(course), MESSAGE_CONSTRAINTS);
+        checkArgument(isValidDate(startDate), DATE_CONSTRAINTS);
+        this.course = course;
+        this.startDate = Optional.of(startDate);
+        this.endDate = Optional.empty();
+    }
 
     /**
-     * Constructs an {@code Course}.
+     * Construct a {@code Course} with start date and end date
```
Comment (xenosf) [inline src/main/java/networkbook/model/person/Course.java:60]:
```suggestion
     * Constructs a {@code Course} with start date and end date.
```

File: src/main/java/networkbook/model/person/Course.java [hunk 3]
```diff
@@ -3,45 +3,145 @@
...
+                    .ofPattern("dd-MM-uuuu")
+                    .withResolverStyle(ResolverStyle.STRICT);
+            return dtf.format(dtf.parse(date)).equals(date);
+        } catch (DateTimeParseException | NullPointerException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Return true if the second given string corresponds with a date that is later than the first date.
```
Comment (xenosf) [inline src/main/java/networkbook/model/person/Course.java:112]:
```suggestion
     * Returns true if the second given string corresponds with a date that is later than the first date.
```

File: src/main/java/networkbook/model/person/Course.java [hunk 4]
```diff
@@ -3,45 +3,145 @@
...
     public String toString() {
-        return value;
+        StringBuilder tsb = new StringBuilder(course);
+        String start = startDateExists() ? " (Started: " + startDate.get() + ")" : "";
+        String end = endDateExists() ? " (Ended: " + endDate.get() + ")" : "";
+        return tsb.append(start).append(end).toString();
     }
 
+    /**
+     * Return true if two Courses have the same 'course' field
```
Comment (xenosf) [inline src/main/java/networkbook/model/person/Course.java:140]:
```suggestion
     * Returns true if two Courses have the same 'course' field.
```

File: src/main/java/networkbook/model/person/Course.java [hunk 5]
```diff
@@ -3,45 +3,145 @@
...
 
-    /*
-     * The first character of the course must not be a whitespace,
-     * otherwise " " (a blank string) becomes a valid input.
+    private final String course;
+    private final Optional<String> startDate;
+    private final Optional<String> endDate;
+
+    /**
+     * Construct a {@code Course}.
```
Comment (xenosf) [inline src/main/java/networkbook/model/person/Course.java:31]:
```suggestion
     * Constructs a {@code Course}.
```

File: src/main/java/networkbook/model/person/Course.java [hunk 6]
```diff
@@ -3,45 +3,145 @@
...
+        requireNonNull(course);
+        checkArgument(isValidCourse(course), MESSAGE_CONSTRAINTS);
+        this.course = course;
+        startDate = Optional.empty();
+        endDate = Optional.empty();
+    }
 
-    public final String value;
+    /**
+     * Construct a {@code Course} with start date
```
Comment (xenosf) [inline src/main/java/networkbook/model/person/Course.java:44]:
```suggestion
     * Constructs a {@code Course} with start date.
```

**Review by xenosf — CHANGES_REQUESTED** [review]
Many of the header comments are not following the [convention](https://se-education.org/guides/conventions/java/index.html#comments), specifically the first sentence. Other than that, LGTM

**Comment by nknguyenhc** [comment]
Merged with [PR#95](https://github.com/AY2324S1-CS2103T-T08-2/tp/pull/95).

**Pull request closed without merging**